### PR TITLE
CB-6341 Implement Redbeams start-stop functionality

### PR DESCRIPTION
**Pull request by pkedvessy**
CB-6341 Implement Redbeams start-stop functionality

### CODE REVIEW

File: cloud-aws/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/connector/resource/AwsRdsStopService.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    private AwsPollTaskFactory awsPollTaskFactory;
+
+    @Inject
+    private AwsBackoffSyncPollingScheduler<Boolean> awsBackoffSyncPollingScheduler;
+
+    public void stop(AuthenticatedContext ac, String dbInstanceIdentifier) throws ExecutionException, TimeoutException, InterruptedException {
+        AwsCredentialView credentialView = new AwsCredentialView(ac.getCloudCredential());
+        String regionName = ac.getCloudContext().getLocation().getRegion().value();
+        AmazonRDS rdsClient = awsClient.createRdsClient(credentialView, regionName);
+
```
Comment (doktoric) [inline cloud-aws/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/connector/resource/AwsRdsStopService.java:36]: some log would be nice that we would like stop the which db instance

File: cloud-aws/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/connector/resource/AwsRdsStartService.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    @Inject
+    private AwsPollTaskFactory awsPollTaskFactory;
+
+    @Inject
+    private AwsBackoffSyncPollingScheduler<Boolean> awsBackoffSyncPollingScheduler;
+
+    public void start(AuthenticatedContext ac, String dbInstanceIdentifier) throws ExecutionException, TimeoutException, InterruptedException {
+        AwsCredentialView credentialView = new AwsCredentialView(ac.getCloudCredential());
+        String regionName = ac.getCloudContext().getLocation().getRegion().value();
+        AmazonRDS rdsClient = awsClient.createRdsClient(credentialView, regionName);
```
Comment (doktoric) [inline cloud-aws/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/connector/resource/AwsRdsStartService.java:35]: some log would be nice that we would like start the which db instance

File: redbeams/src/main/java/com/sequenceiq/redbeams/flow/redbeams/start/handler/StartDatabaseServerHandler.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    @Override
+    public void accept(Event<StartDatabaseServerRequest> event) {
+        LOGGER.debug("Received event: {}", event);
+        StartDatabaseServerRequest request = event.getData();
+        CloudContext cloudContext = request.getCloudContext();
+        try {
+            CloudConnector<Object> connector = cloudPlatformConnectors.get(cloudContext.getPlatformVariant());
+            AuthenticatedContext ac = connector.authentication().authenticate(cloudContext, request.getCloudCredential());
+
+            connector.resources().startDatabaseServer(ac, request.getDbInstanceIdentifier());
```
Comment (doktoric) [inline redbeams/src/main/java/com/sequenceiq/redbeams/flow/redbeams/start/handler/StartDatabaseServerHandler.java:48]: I think we should check first that do we really need to start the initiate start action or not.
When one of the CB node dies and trying to restart the flow we dont want to initiate the start event twice

File: redbeams/src/main/java/com/sequenceiq/redbeams/flow/redbeams/stop/handler/StopDatabaseServerHandler.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+    @Override
+    public void accept(Event<StopDatabaseServerRequest> event) {
+        LOGGER.debug("Received event: {}", event);
+        StopDatabaseServerRequest request = event.getData();
+        CloudContext cloudContext = request.getCloudContext();
+        try {
+            CloudConnector<Object> connector = cloudPlatformConnectors.get(cloudContext.getPlatformVariant());
+            AuthenticatedContext ac = connector.authentication().authenticate(cloudContext, request.getCloudCredential());
+
+            connector.resources().stopDatabaseServer(ac, request.getDbInstanceIdentifier());
```
Comment (doktoric) [inline redbeams/src/main/java/com/sequenceiq/redbeams/flow/redbeams/stop/handler/StopDatabaseServerHandler.java:57]: I think we should check first that do we really need to start the initiate stop action or not.
When one of the CB node dies and trying to restart the flow we dont want to initiate the stop event twice

Comment (pkedvessy) [reply redbeams/src/main/java/com/sequenceiq/redbeams/flow/redbeams/stop/handler/StopDatabaseServerHandler.java:57]: Its already done in https://github.com/hortonworks/cloudbreak/blob/fbe8818ef58b368634efdc3e7e16b282435014b9/redbeams/src/main/java/com/sequenceiq/redbeams/service/stack/RedbeamsStopService.java#L43

Should I implement it on AWS level as well?

File: redbeams/src/main/java/com/sequenceiq/redbeams/flow/redbeams/start/handler/StartDatabaseServerHandler.java (hunk 3 above)
Comment (pkedvessy) [reply redbeams/src/main/java/com/sequenceiq/redbeams/flow/redbeams/start/handler/StartDatabaseServerHandler.java:48]: Its already done in
https://github.com/hortonworks/cloudbreak/blob/fbe8818ef58b368634efdc3e7e16b282435014b9/redbeams/src/main/java/com/sequenceiq/redbeams/service/stack/RedbeamsStartService.java#L43
Should I implement it on AWS level as well?

File: cloud-aws/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/connector/resource/AwsRdsStartService.java (hunk 2 above)
Comment (pkedvessy) [reply cloud-aws/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/connector/resource/AwsRdsStartService.java:35]: Done in https://github.com/hortonworks/cloudbreak/blob/fbe8818ef58b368634efdc3e7e16b282435014b9/redbeams/src/main/java/com/sequenceiq/redbeams/service/stack/RedbeamsStartService.java#L41

Should I implement it here as well?

File: cloud-aws/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/connector/resource/AwsRdsStopService.java (hunk 1 above)
Comment (pkedvessy) [reply cloud-aws/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/connector/resource/AwsRdsStopService.java:36]: Done in https://github.com/hortonworks/cloudbreak/blob/fbe8818ef58b368634efdc3e7e16b282435014b9/redbeams/src/main/java/com/sequenceiq/redbeams/service/stack/RedbeamsStopService.java#L41

Should I implement it here as well?

Comment (doktoric) [reply cloud-aws/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/connector/resource/AwsRdsStopService.java:36]: nope thats cool thanks

File: redbeams/src/main/java/com/sequenceiq/redbeams/flow/redbeams/start/handler/StartDatabaseServerHandler.java (hunk 3 above)
Comment (doktoric) [reply redbeams/src/main/java/com/sequenceiq/redbeams/flow/redbeams/start/handler/StartDatabaseServerHandler.java:48]: not on aws level but in the abstract flow step

File: redbeams/src/main/java/com/sequenceiq/redbeams/flow/redbeams/stop/handler/StopDatabaseServerHandler.java (hunk 4 above)
Comment (doktoric) [reply redbeams/src/main/java/com/sequenceiq/redbeams/flow/redbeams/stop/handler/StopDatabaseServerHandler.java:57]: not on aws level but in the abstract flow step

Comment (pkedvessy) [reply redbeams/src/main/java/com/sequenceiq/redbeams/flow/redbeams/stop/handler/StopDatabaseServerHandler.java]: Done.

**Comment by pkedvessy** [comment]
> As I cannot comment on unchanged files:
> Please consider changing RedbeamsFlowInformation and add the new flows to the restartable flows.
@gergopapi2 
Done, thanks for checking the PR.

File: redbeams/src/main/java/com/sequenceiq/redbeams/flow/redbeams/start/actions/AbstractRedbeamsStartAction.java [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+import javax.annotation.PostConstruct;
+import javax.inject.Inject;
+import java.util.Map;
+import java.util.Optional;
+
+import static com.sequenceiq.cloudbreak.cloud.model.AvailabilityZone.availabilityZone;
+import static com.sequenceiq.cloudbreak.cloud.model.Location.location;
+import static com.sequenceiq.cloudbreak.cloud.model.Region.region;
+
+public abstract class AbstractRedbeamsStartAction<P extends Payload>
```
Comment (gergopapi2) [inline redbeams/src/main/java/com/sequenceiq/redbeams/flow/redbeams/start/actions/AbstractRedbeamsStartAction.java:30]: AbstractRedbeamsStartAction and AbstractRedbeamsStopAction are very similar, are almost the same. Also, they are quite similar to AbstractRedbeamsProvisionAction and TerminationAction. I would try to create one common class from them, AbstractRedbeamsAction and use it in all four flows.

If that is too big of a task (~more than an 1-2 hours of work), then I would create a separate refactor jira for it.

Comment (pkedvessy) [reply redbeams/src/main/java/com/sequenceiq/redbeams/flow/redbeams/start/actions/AbstractRedbeamsStartAction.java:30]: If you don't mind, I'd more prefer the refactor as a separated task.

File: redbeams/src/main/java/com/sequenceiq/redbeams/service/stack/RedbeamsStartService.java [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    private DBStackService dbStackService;
+
+    @Inject
+    private DBStackStatusUpdater dbStackStatusUpdater;
+
+    @Inject
+    private RedbeamsFlowManager flowManager;
+
+    public void startDatabaseServer(String crn) {
+        DatabaseServerConfig databaseServerConfig = databaseServerConfigService.getByCrn(crn);
```
[the remark this answers is not in the document]
Comment (pkedvessy) [reply redbeams/src/main/java/com/sequenceiq/redbeams/service/stack/RedbeamsStartService.java]: Done. Done in the RedbeamsStopService as well.

[the remark this answers is not in the document]
Comment (pkedvessy) [reply redbeams/src/main/java/com/sequenceiq/redbeams/service/stack/RedbeamsStartService.java]: Done. Done in the RedbeamsStopService as well.

**Pull request merged**